ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,501 @@
+          AccountInfoMap newAccountInfoMap = new AccountInfoMap(remoteAccountMap);
+          accountInfoMapRef.set(newAccountInfoMap);
+          logger.trace("Completed updating local cache with number={} accounts",
+              newAccountInfoMap.getAccounts().size());
+        }
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Checks if there are duplicate accountId or accountName in the given collection of {@link Account}s.
+   *
+   * @param accounts A collection of {@link Account}s to check duplication.
+   * @return {@code true} if there are duplicated accounts in id or name in the given collection of accounts,
+   *                      {@code false} otherwise.
+   */
+  private boolean hasDuplicateAccountIdOrName(Collection<Account> accounts) {
+    if (accounts == null) {
+      return false;
+    }
+    Set<Short> idSet = new HashSet<>();
+    Set<String> nameSet = new HashSet<>();
+    boolean res = false;
+    for (Account account : accounts) {
+      if (!idSet.add(account.getId()) || !nameSet.add(account.getName())) {
+        logger.debug("Accounts to update have conflicting id or name. Conflicting accountId={} accountName={}",
+            account.getId(), account.getName());
+        res = true;
+        break;
+      }
+    }
+    return res;
+  }
+
+  /**
+   * Checks a collection of {@link Account}s if there is any conflict between the {@link Account}s to update and the
+   * {@link Account}s in the local cache.
+   *
+   * @param accountsToSet The collection of {@link Account}s to check with the local cache.
+   * @return {@code true} if there is at least one {@link Account} in {@code accounts} conflicts with the {@link Account}s
+   *                      in the cache, {@code false} otherwise.
+   */
+  private boolean hasConflictWithCache(Collection<Account> accountsToSet) {
+    return hasConflictingAccount(accountsToSet, accountInfoMapRef.get());
+  }
+
+  /**
+   * Checks if there is any {@link Account} in a given collection of {@link Account}s conflicts against any {@link Account}
+   * in a {@link AccountInfoMap}, according to the Javadoc of {@link AccountService}. Two {@link Account}s can be
+   * conflicting with each other if they have different account Ids but the same account name.
+   *
+   * @param accountsToSet The collection of {@link Account}s to check conflict.
+   * @param currentAccounts A {@link AccountInfoMap} that represents a group of {@link Account}s to check conflict.
+   * @return {@code true} if there is at least one {@link Account} in {@code accountPairs} conflicts with the existing
+   *                      {@link Account}s in {@link AccountInfoMap}, {@code false} otherwise.
+   */
+  private boolean hasConflictingAccount(Collection<Account> accountsToSet, AccountInfoMap currentAccounts) {
+    boolean res = false;
+    for (Account account : accountsToSet) {
+      Account remoteAccount = currentAccounts.getAccountById(account.getId());
```

Comment:
I do not have a good reason for naming it as `remoteAccount`, though it is a remote account. But in the method's scope, this is not the reason. I have renamed it to `accountInMap`, though it might still be a bad name.